xarray/backends/rasterio_.py

```diff
@@ -0,0 +1,120 @@
+import numpy as np
+
+try:
+    import rasterio
+except ImportError:
+    rasterio = False
+
+from .. import Variable, DataArray
+from ..core.utils import FrozenOrderedDict, Frozen, NDArrayMixin
+from ..core import indexing
+from ..core.pycompat import OrderedDict, suppress
+
+from .common import AbstractDataStore
+
+__rio_varname__ = 'raster'
+
+
+class RasterioArrayWrapper(NDArrayMixin):
+    def __init__(self, ds):
+        self._ds = ds
+        self.array = ds.read()
```

Comment:
@NicWayand -

I think you can basically drop this in replacing the current `RasterioArrayWrapper`:

```Python
class RasterioArrayWrapper(NdimSizeLenMixin):
    """Mixin class RasterIO datasets for making wrappers of N-dimensional 
    arrays that conform to the ndarray interface required for the data
    argument to Variable objects.
    A subclass should set the `array` property and override one or more of
    `dtype`, `shape` and `__getitem__`.
    """
    def __init__(self, ds):
        self.ds = ds
    
    @property
    def dtype(self):
        if len(set(self.ds.dtypes[0])) != 1:
            raise ValueError(
                'Can only handle Rastio dataset with all bands having the same type')
        return np.dtype(self.ds.dtypes[0])

    @property
    def shape(self):
        return self.ds.shape

    def __array__(self, dtype=None):
        '''Not sure if this will work as is'''
        return np.asarray(self[...], dtype=dtype)

    def __getitem__(self, key):
        band = range(self.shape[0])[key[0]]
        window = []
        for win in key[1:]:
            if instance(win, slice):
                window.append((win.start, win.stop))
            elif isinstance(win, int):
                window.append((win, win + 1))
            else:  # integer ndarray
                window.append((win.min(), win.max()))
        raw_data = self.ds.read(band, window=window)
        # now, fix up raw_data to conform to numpy indexing conventions
        # - drop axes for integer band/windows
        # - stride if windows are slices with win.step != 1
        # - subset if window is an integer ndarray

    def __repr__(self):
        return '%s(array=%r)' % (type(self).__name__, self.ds)
```

I'm not sure the `__array__` will work as it is until we get the `__get_item__` fully functional.